### PR TITLE
Allow start/pause of streaming image input. Only access the webcam while it's needed

### PR DESCRIPTION
**Pull request by freddyaboulton**
## Description

Closes: #3642
Closes: #7230

If the webcam is an allowed image source, gradio will access the webcam as soon as the component is mounted on the page. There is also no way to stop gradio from accessing the webcam. This is #3642

When the webcam is an allowed source, users have to click into the image component to grant webcam access. This is what it looks like in that case:

image

After the webcam stops recording or taking a picture, gradio will stop using the webcam and revert back to the ^previous state.

Additionally, when `gr.Image(streaming=true)`, the webcam component will behave similarly to video recording. There will be a record and stop button at the bottom that users can use to stop the stream of images. 

### Webcam Demo no streaming
webcam_demo

### Webcam Demo with streaming
stream_frames

Test streaming with this demo
```python
import gradio as gr
import numpy as np

def flip(im):
    return np.flipud(im)

with gr.Blocks() as demo:
    with gr.Row():
        input_img = gr.Image(sources=["webcam"])
        output_img = gr.Image()
        input_img.stream(flip, input_img, output_img)
demo.launch(share=True)

if __name__ == "__main__":
    demo.launch()
    
```

EDIT:

Also fixes Interface streaming blocking the output, fixes: #7021 

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

### CODE REVIEW

File: js/image/shared/Webcam.svelte [hunk 1]
```diff
@@ -180,19 +203,34 @@
 		event.stopPropagation();
 		options_open = false;
 	}
+
+	$: console.log("webcam_accessed", webcam_accessed);
```
Comment (freddyaboulton) [inline js/image/shared/Webcam.svelte:207]: will delete 🤦

File: js/image/shared/Webcam.svelte [hunk 2]
```diff
@@ -180,19 +203,34 @@
...
+		class:flip={mirror_webcam}
+		class:invisible={!webcam_accessed}
+	/>
+	{#if !webcam_accessed}
+		<div class="wrap-webcam-access">
+			<span class="icon-wrap">
+				<WebcamIcon />
+			</span>
+			<button on:click={async () => await access_webcam()}
+				>{"Click to access webcam"}</button
```
Comment (freddyaboulton) [inline js/image/shared/Webcam.svelte:222]: This should probably be camel case for consistency

**Comment by aliabid94** [comment]
I could've sworn I made the change to set the progress tracker visibility to False when streaming from a webcam

**Comment by abidlabs** [comment]
> I could've sworn I made the change to set the progress tracker visibility to False when streaming from a webcam

I'm pretty sure its a regression. We should do this for any streaming event, e.g. https://github.com/gradio-app/gradio/issues/7230 as well

**Comment by abidlabs** [comment]
This looks great to me @freddyaboulton! My only feedback is that the clickable area to give access to the webcam is very small. I kept clicking the icon and was confused why it wasn't working. I'd suggest making this whole area clickable:

image

**Comment by aliabid94** [comment]
Is the demo supposed to be using an audio component? Btw when I was running locally I did not see the progress bar but also I got this error after a while when streaming a webcam:
`ValueError: Cannot process this value as an Image, it is of type: <class 'tuple'>`

**Comment by aliabid94** [comment]
> Btw when I was running locally I did not see the progress bar

Actually there was a visible progress bar only when using streaming from Interface, fixed in this PR in this commit: https://github.com/gradio-app/gradio/pull/7228/commits/db019e4c2a61f8fe16f68dc915982c060182267d

**Comment by freddyaboulton** [comment]
> Actually there was a visible progress bar only when using streaming from Interface, fixed in this PR in this commit: https://github.com/gradio-app/gradio/commit/db019e4c2a61f8fe16f68dc915982c060182267d

Nice - confirmed it works! 🔥 

> Is the demo supposed to be using an audio component?

Yea that's my bad - copy/pasted the wrong demo. I updated the PR description to include the correct one but also including it here

```python
import gradio as gr
import numpy as np

def flip(im):
    return np.flipud(im)

with gr.Blocks() as demo:
    with gr.Row():
        input_img = gr.Image(sources=["webcam"])
        output_img = gr.Image()
        input_img.stream(flip, input_img, output_img)
demo.launch(share=True)

if __name__ == "__main__":
    demo.launch()
```

> My only feedback is that the clickable area to give access to the webcam is very small. 

Agreed - just pushed up a fix! Will wait for CI to pass before opening it up for broader review.

File: js/image/shared/Webcam.svelte [hunk 3]
```diff
@@ -131,7 +142,20 @@
...
 
-	access_webcam();
+	let webcam_accessed = false;
+
+	function record_video_or_photo(): void {
+		if (mode === "image" && streaming) {
+			recording = !recording;
+		}
+		const func = mode === "image" ? take_picture : take_recording;
+		func();
```
Comment (aliabid94) [inline js/image/shared/Webcam.svelte:152]: let's avoid abbreviations like `func`. Personally think this is more readable tbh
```suggestion
		if (mode === "image") {
		  take_picture();
		} else {
		  take_recording();
		}
```

File: js/image/shared/Webcam.svelte [hunk 4]
```diff
@@ -253,6 +282,10 @@
 		height: var(--size-full);
 	}
 
+	.invisible {
```
Comment (aliabid94) [inline js/image/shared/Webcam.svelte:285]: nit, but we use the class name `hide` throughout the codebase for `display: none;`, for consistency would be nice

**Review by aliabid94 — APPROVED** [review]
LGTM great work

**Comment by dawoodkhan82** [comment]
Looks great @freddyaboulton. Small nit, I notice after taking a picture the input component quickly flashes back to the "click to access webcam" state before showing the taken picture.

**Comment by abidlabs** [comment]
Maybe also fixes: https://github.com/gradio-app/gradio/issues/7230?

**Comment by freddyaboulton** [comment]
Thanks for the reviews everyone! Will merge the comments and flicker have been addressed.

**Pull request merged**